lib/Sema/TypeCheckType.cpp

```diff
@@ -2148,6 +2148,23 @@ Type TypeResolver::resolveAttributedType(TypeAttributes &attrs,
   if (!ty) ty = resolveType(repr, instanceOptions);
   if (!ty || ty->hasError()) return ty;
 
+  // We may be inside a protocol and as a result, the resolved type
+  // ends up being a DependentMemberType, pointing to a type alias
+  // inside the protocol. Look up the type alias inside the protocol
+  // and get the underlying type.
+  if (auto DMT = dyn_cast<DependentMemberType>(ty.getPointer())) {
```

Comment:
I would instead just say that if there is a structural resolver we ignore the attributes

Reply:
@slavapestov do you mean something like

```c++
if (auto DMT = dyn_cast<DependentMemberType>(ty.getPointer())) {
    if (resolution.getStage() == TypeResolutionStage::Structural) {
      attrs.clearAttribute(TAK_escaping);
    }
  }
```

Reply:
Yes, just ignore all attributes in this case, but do test that the attribute is eventually applied, by calling the requirement with a noescape function and expecting a diagnostic

Reply:
I think @slavapestov means:

```c++
if (ty->is<DependentMemberType>() && 
    resolution.getStage() == TypeResolutionStage::Structural) {
    return ty;
}
```

Reply:
Also it looks like we'd need to do this for `@autoclosure` too:

```
protocol P {
  typealias F = (Int) -> Void
  func foo<T>(_: T, _: @autoclosure F)
}
```

produces `error: @autoclosure attribute only applies to function types`

Reply:
But i guess it makes more sense to move @autoclosure diagnostics to `validateParameterType` because this attribute is associated with declaration now.